win32/win32.c

```diff
@@ -7257,97 +7241,131 @@ rb_w32_read_internal(int fd, void *buf, size_t size, rb_off_t *offset)
     }
 
     ret = 0;
-    isconsole = is_console(_osfhnd(fd)) && (osver.dwMajorVersion < 6 || (osver.dwMajorVersion == 6 && osver.dwMinorVersion < 2));
-    if (isconsole) {
-        DWORD mode;
-        GetConsoleMode((HANDLE)_osfhnd(fd),&mode);
-        islineinput = (mode & ENABLE_LINE_INPUT) != 0;
-    }
+    isconsole = is_console(_osfhnd(fd));
   retry:
-    /* get rid of console reading bug */
-    if (isconsole) {
-        constat_reset((HANDLE)_osfhnd(fd));
-        if (start)
-            len = 1;
-        else {
-            len = 0;
-            start = 1;
-        }
-    }
-    else
-        len = size;
+    len = size;
     size -= len;
 
-    if (setup_overlapped(&ol, fd, FALSE, offset)) {
-        rb_acrt_lowio_unlock_fh(fd);
-        return -1;
-    }
+    if (isconsole) {
+        /* Direct console input, without pipe */
```

Comment:
@nobu, @unak Can I get some help from you here?

`ReadConsoleW` retrieves console characters as UTF-16. I think the most useful way to pass the input data to ruby API is as UTF-8 encoded string. There seems to be no easy way to to archive this through the Win32-API.

The below implementation requires to `read` at least so many bytes that the UTF-8 converted string fits into it. Otherwise `STDIN.read` returns `nil`. In contrast `STDIN.gets` provides a large buffer size so that the string is converted to UTF-8 and returned.

To properly implement reading UTF-8 characters bytewise some internal state has to be managed, about the character position of already delivered UTF-8 bytes and about pending/buffered bytes. Can you recommend a way to store this information?

Maybe the `socklist` hash table could be extended to store this data. But maybe there is some better place. Or do you have some other idea how to implement Unicode aware console reading?